### PR TITLE
New differential functionality

### PR DESCRIPTION
**Pull request by samjrholt**

### CODE REVIEW

File: discretisedfield/field.py [hunk 1]
```diff
@@ -2125,7 +2142,7 @@ def div(self):
 
         return sum(
             [
-                getattr(self, self.vdims[i]).derivative(dfu.raxesdict[i])
+                getattr(self, self.vdims[i]).diff(dfu.raxesdict[i])
                 for i in range(self.nvdim)
             ]
         )
```
Comment (lang-m) [inline discretisedfield/field.py:2148]: The extra square brackets can be removed.

File: discretisedfield/field.py [hunk 2]
```diff
@@ -2055,7 +2072,7 @@ def grad(self):
             msg = f"Cannot compute gradient for nvdim={self.nvdim} field."
             raise ValueError(msg)
 
-        return self.derivative("x") << self.derivative("y") << self.derivative("z")
+        return self.diff("x") << self.diff("y") << self.diff("z")
 
     @property
     def div(self):
```
Comment (lang-m) [inline discretisedfield/field.py:2078]: The double backslashes in the docstring have to be changed to single backslashes (because it is a raw string). Probably also in the other docstrings.

File: discretisedfield/field.py [hunk 3]
```diff
@@ -2125,7 +2142,7 @@ def div(self):
 
         return sum(
             [
-                getattr(self, self.vdims[i]).derivative(dfu.raxesdict[i])
+                getattr(self, self.vdims[i]).diff(dfu.raxesdict[i])
                 for i in range(self.nvdim)
```
Comment (lang-m) [inline discretisedfield/field.py:2146]:
```suggestion
                getattr(self, vdim).diff(vdim) for vdim in self.nvdim
```

File: discretisedfield/tests/test_field.py [hunk 4]
```diff
@@ -1218,16 +1228,63 @@ def value_fun(point):
...
+        cell = (2, 2, 2)
+
+        mesh_nodirichlet = df.Mesh(p1=p1, p2=p2, cell=cell)
+        mesh_dirichlet = df.Mesh(p1=p1, p2=p2, cell=cell, bc="dirichlet")
+
+        # Scalar field
+        def value_fun(point):
+            return point[0] * point[1] * point[2]
+
+        # No Neumann
```
Comment (lang-m) [inline discretisedfield/tests/test_field.py:1265]: No Dirichlet?

File: discretisedfield/tests/test_field.py [hunk 5]
```diff
@@ -1218,16 +1228,63 @@ def value_fun(point):
...
+        f1 = df.Field(mesh_nodirichlet, nvdim=1, value=value_fun)
+        assert np.isclose(f1.diff("x")((11, 1, 1)), 1)
+        assert np.isclose(f1.diff("y")((1, 7, 1)), 1)
+        assert np.isclose(f1.diff("z")((1, 1, 5)), 1)
+
+        # Neumann
```
Comment (lang-m) [inline discretisedfield/tests/test_field.py:1271]: Dirichlet?

**Review by lang-m — CHANGES_REQUESTED** [review]
The changes in the code look good. I commented on a few minor things.

In the tests we are using `np.isclose` and `np.allclose` inconsistenly. It propably does not matter. Is there any particular reason for doing so.

I did not check all the tests in detail (i.e. compute the derivatives to check if the tests are actually correct). I assume you have checked it?

File: discretisedfield/field.py [hunk 6]
```diff
@@ -1915,66 +1916,82 @@ def derivative(self, direction, n=1):
...
-        if dfu.raxesdict[direction] in self.mesh.bc:  # PBC
-            pad_width = {dfu.raxesdict[direction]: (1, 1)}
+        if direction in self.mesh.bc:  # PBC
+            pad_width = {direction: (1, 1)}
             padding_mode = "wrap"
         elif self.mesh.bc == "neumann":
-            pad_width = {dfu.raxesdict[direction]: (1, 1)}
-            padding_mode = "edge"
+            pad_width = {direction: (1, 1)}
+            padding_mode = "symmetric"
```
[the remark this answers is not in the document]
Comment (samjrholt) [reply discretisedfield/field.py:1934]: You are correct, they are equivalent for a single cell but not for more.
`edge` : `abcde|eee`
`symmetric` : `abcde|edc`

**Comment by samjrholt** [comment]
> The changes in the code look good. I commented on a few minor things.
> 
> In the tests we are using `np.isclose` and `np.allclose` inconsistenly. It propably does not matter. Is there any particular reason for doing so.
> 
> I did not check all the tests in detail (i.e. compute the derivatives to check if the tests are actually correct). I assume you have checked it?

I have gone through and changed `np.isclose` to `np.allclose` for consistency.
I have also checked the values for the tests given the correct boundary conditions. (And now added those that were missing)

**Review by lang-m — APPROVED** [review]
Looks good to me. You can merge.
